[BUGFIX] Fix hang and TypeError when invoking `database:import` through the symfony/console PHP API

adamkoppede opened the pull request:
Fix a hang (indefinite wait) and a PHP runtime TypeError that occurred when invoking `database:import` via PHP-code inside another TYPO3 console command with `ArrayInput`.

[A reproduction example is available in another github project](https://github.com/adamkoppede/typo3-console-in-process-command-test-cases) because I don't know how to integrate it into this project's testing framework.

adamkoppede committed:
[BUGFIX] Wrong input stream used in database:import called via PHP-API

The "database:import" console command used the wrong input stream when
it is used with an input stream that isn't STDIN. The following concrete
example hangs:

    $commandRegistry = GeneralUtility::makeInstance(CommandRegistry::class);
    $command = $commandRegistry->get('database:import');

    $input = new ArrayInput(['--connection' => 'Default']);
    $inputStream = fopen('/some.sql', 'b+');
    assert($inputStream !== false);
    $input->setStream($inputStream);

    $exitCode = $command->run($input, new ConsoleOutput());
    assert($exitCode === Command::SUCCESS);
    assert(fclose($inputStream));

A full reproduction of the issue is available at
https://github.com/adamkoppede/typo3-console-in-process-command-test-cases.

See: https://github.com/symfony/symfony/issues/37835#issuecomment-674386588

adamkoppede committed:
[BUGFIX] Fix TypeError in MysqlCommand::__construct()

When running the "database:import" console command via the
symfony/console PHP API with an output implementation that isn't
ConsoleOutput, the following PHP runtime TypeError occurred:

    TypeError:
        Helhum\Typo3Console\Database\Process\MysqlCommand::__construct():
            Argument #2 ($output) must be of type ?Symfony\Component\Console\Output\ConsoleOutput,
            Symfony\Component\Console\Output\StreamOutput given,
            called in vendor/helhum/typo3-console/Classes/Console/Command/Database/DatabaseImportCommand.php on line 89
            and defined in vendor/helhum/typo3-console/Classes/Console/Database/Process/MysqlCommand.php:38
    Stack trace:
        #0 vendor/helhum/typo3-console/Classes/Console/Command/Database/DatabaseImportCommand.php(89): Helhum\Typo3Console\Database\Process\MysqlCommand->__construct()
        #1 vendor/symfony/console/Command/Command.php(298): Helhum\Typo3Console\Command\Database\DatabaseImportCommand->execute()
        #2 Classes/TestCommand.php(68): Symfony\Component\Console\Command\Command->run()

A full reproduction of the issue is available at
https://github.com/adamkoppede/typo3-console-in-process-command-test-cases.

OutputInterface::OUTPUT_RAW is required to avoid going through the
output formatter in Output::write().

helhum commented on Classes/Console/Database/Process/MysqlCommand.php:
```diff
@@ -89,9 +91,12 @@ private function buildDefaultOutputCallback($outputCallback): callable
             $outputCallback = function ($type, $data) {
                 if (Process::OUT === $type) {
-                    echo $data;
+                    $this->output->write($data);
```
This is not feasible imho, because by default this runs through a formatter. Back then I experimented with writing to output without formatters, but never got it as reliable as simply echoing the data out. Disabling formatting is the minimum that needs to be changed

adamkoppede replied:
You're right, I didn't notice that _OutputInterface_ is concerted with Formatting. I thought that this is what _SymfonyStyle_ is for.

[From just reading the implementation of Output::write()](https://github.com/symfony/console/blob/v7.2.1/Output/Output.php#L108-L138), _OutputInterface::OUTPUT\_RAW_ seams to be the option we're looking for here.

adamkoppede replied:
I was able to reproduce the undesired formatting using the following command input payload: `$sql = 'SELECT "<error>I shouldn\'t be styled because I\'m user data</error>";';`. Great Catch!

Adding `OutputInterface::OUTPUT_RAW` fixed that case. I'm adding a test case in _DatabaseCommandControllerTest.php_.

helhum replied:
I re-read the code path with `OutputInterface::OUTPUT_RAW` and it seems fine indeed. Let's give this a try...

helhum reviewed (changes requested):
Input Stream changes are totally fine. Output changes need more care I think

adamkoppede committed:
[TASK] Add test case for symfony formatting instructions in SQL output

Symfony formatting instructions like `<error>text</error>` that may be
contained in the MysqlCommand output must be threaded as opaque
arbitrary data and mustn't be interpreted by any OutputFormatter.

See: https://github.com/TYPO3-Console/TYPO3-Console/pull/1202#discussion_r1961868291

helhum reviewed (dismissed):
I'll do some manual tests before merge. Thanks so far!
